pkg/coscheduling/coscheduling.go

```diff
@@ -95,294 +83,364 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
-	return nil
+	return cs
 }
 
-// Permit is the functions invoked by the framework at "Permit" extension point.
-func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
-	pgInfo, _ := cs.getOrCreatePodGroupInfo(pod, time.Now())
-	if len(pgInfo.key) == 0 {
-		return framework.NewStatus(framework.Success, ""), 0
-	}
+// AddPod is called by the framework while trying to evaluate the impact
+// of adding podToAdd to the node while scheduling podToSchedule.
+func (cs *Coscheduling) AddPod(ctx context.Context, _ *framework.CycleState,
+	_ *v1.Pod, podToAdd *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	return framework.NewStatus(framework.Success, "")
+}
 
-	namespace := pod.Namespace
-	podGroupName := pgInfo.name
-	minAvailable := pgInfo.minAvailable
-	// bound includes both assigned and assumed Pods.
-	bound := cs.calculateBoundPods(podGroupName, namespace)
-	// The bound is calculated from the snapshot. The current pod does not exist in the snapshot during this scheduling cycle.
-	current := bound + 1
-
-	if current < minAvailable {
-		klog.V(3).Infof("The count of podGroup %v/%v/%v is not up to minAvailable(%d) in Permit: current(%d)",
-			pod.Namespace, podGroupName, pod.Name, minAvailable, current)
-		// TODO Change the timeout to a dynamic value depending on the size of the `PodGroup`
-		return framework.NewStatus(framework.Wait, ""), time.Duration(*cs.args.PermitWaitingTimeSeconds) * time.Second
+// RemovePod is called by the framework while trying to evaluate the impact
+// of removing podToRemove from the node while scheduling podToSchedule.
+func (cs *Coscheduling) RemovePod(ctx context.Context, _ *framework.CycleState,
+	podToSchedule *v1.Pod, podToRemove *v1.Pod, _ *framework.NodeInfo) *framework.Status {
+	if err := cs.pgMgr.PreemptRemovePod(ctx, podToSchedule, podToRemove); err != nil {
+		if klog.V(5).Enabled() {
+			klog.Error(err)
+		}
+		return framework.NewStatus(framework.Unschedulable, err.Error())
 	}
+	klog.V(5).Infof("CoschedulingExtension pass remove pod %v", podToRemove.Name)
+	return framework.NewStatus(framework.Success, "")
+}
 
-	klog.V(3).Infof("The count of PodGroup %v/%v/%v is up to minAvailable(%d) in Permit: current(%d)",
-		pod.Namespace, podGroupName, pod.Name, minAvailable, current)
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
-			waitingPod.Allow(cs.Name())
+// Permit is the functions invoked by the framework at "Permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	var pgs *schecache.PodGroupSchedulingStatus
```

Comment:
pgs -> pgStatus (applies elsewhere)

and move it to the place that is close to its usage (L212).